utilities/include/vec4.hpp

```diff
@@ -0,0 +1,149 @@
+        }
+        
+        constexpr T norm(const bool alpha=ALPHA_) const{ //__GNUC__
+            return sqrt(dot(*this,alpha));
+        }
+
+        constexpr static T angle(const vec4& v1,const vec4& v2,const bool alpha=ALPHA_){//__GNUC__
+            return acos(v1.dot(v2,alpha)/(v1.norm(alpha)*v2.norm(alpha)));
+        }
+        constexpr vec4 unit() const{//__GNUC__
+            return *this/this->norm();
+        }
+
+        
+        constexpr vec4 cross(const vec4& v2) const{ //ignoring the "alpha" channel
+            return vec4(m_v[1] * v2.m_v[2] - m_v[2] * v2.m_v[1], m_v[2] * v2.m_v[0] - m_v[0] * v2.m_v[2],m_v[0] * v2.m_v[1] - m_v[1] * v2.m_v[0],0);
+        }
+
+        //vector addition
+        friend constexpr vec4 operator+(const vec4& v1, const vec4& v2){
+            return b2v(v2b(v1)+v2b(v2));
+        }
+
+        //vector subtraction
+        friend constexpr vec4 operator-(const vec4& v1, const vec4& v2){
+            return b2v(v2b(v1)-v2b(v2));
+        }
+
+        //scalar post-multiplication
+        friend constexpr vec4 operator*(const vec4& v1, const T& s){
+            return b2v(v2b(v1)*xsimd::batch<T,A>::broadcast(s));
+        }
+
+        //scalar pre-multiplication
+        friend constexpr vec4 operator*(const T& s, const vec4& v1){
+            return v1*s;
+        }
+
+        //scalar post-division
+        friend constexpr vec4 operator/(const vec4& v1, const T& s){
+            return v1*(1/double(s));
+        }
+
+        //addition+assignemnt
+        friend constexpr vec4& operator+=(vec4& v1, const vec4& v2){
+            v1=v1+v2;
+            return v1;
+        }
+        
+        //subtraction+assignemnt
+        friend constexpr vec4& operator-=(vec4& v1, const vec4& v2){
+            v1=v1-v2;
+            return v1;
+        }
+    
+        //scalar multiplication+assignment
+        friend constexpr vec4& operator*=(vec4& v1, const float& s){
+            v1=v1*s;
+            return v1;
+        }
+        
+        //scalar division+assignment
+        friend constexpr vec4& operator/=(vec4& v1, const float& s){
+            v1=v1/s;
+            return v1;
+        }
+
+        //negation operator 
+        friend constexpr vec4 operator-(const vec4& v1){
+            return b2v(-v2b(v1));
+        }
+
+        //indexing operation
+        constexpr T operator[](int i){
+            return m_v[i];
+        }
+        
+        //print the vector
+        friend std::ostream& operator<<(std::ostream& out, const vec4& v) {
+			out <<"("<< v.m_v[0] << "," << v.m_v[1] << "," << v.m_v[2]<<"," << v.m_v[3]<<")";
+			return out;
+		}
+
+	private:
+        T m_v[4];
+
+        static constexpr xsimd::batch<T,A> v2b(const vec4& v){
+            return  xsimd::batch<T,A>::load_aligned(v.m_v);
+        }
+        
+        static constexpr vec4 b2v(xsimd::batch<T,A> x){
+            vec4 v{x};
+            return v;
+        }
+
+        static constexpr T sum(xsimd::batch<T,A> x){
```

Comment:
You don't use this function anywhere. Is this intended?

Reply:
I wanted to use it for encapsulating `xsimd::hadd` to sum over elements in dot product but didn't find a sensible way handling garbage values.

Reply:
I have changed the implementation of the dot product. It now multiplies using batch and sum returns the sum of the first 3 elements. We got a speed up in multiplication (we do them parallel now) but that incurs a overhead of converting the batch back to vector and then accessing the elements to sum. Alternatively SIMD has a `get(std::size_t i)` mentioned [here](https://xsimd.readthedocs.io/en/latest/api/xsimd_batch.html?highlight=inefficient#_CPPv4NK5xsimd5batch3getENSt6size_tE), but they mentioned that it is inefficient. We have to examine what performs better.

Reply:
Converting vectors to batches shouldn't be a massive overhead.

I'm marking this as resolved because you addressed the original concern.